flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2159,138 +2159,267 @@ struct CoordinateOpConversion
-    bool hasSubdimension = hasSubDimensions(cpnTy);
-
-    // if argument 0 is complex, get the real or imaginary part
-    if (fir::isa_complex(cpnTy)) {
-      SmallVector<mlir::Value> offs = {c0};
-      offs.append(std::next(operands.begin()), operands.end());
+    mlir::Type baseObjectTy = coor.getBaseType();
+    mlir::Type objectTy = fir::dyn_cast_ptrOrBoxEleTy(baseObjectTy);
+    assert(objectTy && "fir.coordinate_of expects a reference type");
+
+    // Complex type - basically, extract the real or imaginary part
+    if (fir::isa_complex(objectTy)) {
+      mlir::LLVM::ConstantOp c0 =
+          genConstantIndex(loc, lowerTy().indexType(), rewriter, 0);
+      SmallVector<mlir::Value> offs = {c0, operands[1]};
       mlir::Value gep = genGEP(loc, ty, rewriter, base, offs);
       rewriter.replaceOp(coor, gep);
       return success();
     }
 
-    // if argument 0 is boxed, get the base pointer from the box
-    if (auto boxTy = firTy.dyn_cast<fir::BoxType>()) {
-      // Special case:
-      //   %lenp = len_param_index foo, type<T(foo:i32)...>
-      //   %addr = coordinate_of %box, %lenp
-      if (coor.getNumOperands() == 2) {
-        auto coorPtr = *coor.coor().begin();
-        auto *s = coorPtr.getDefiningOp();
-        if (isa_and_nonnull<fir::LenParamIndexOp>(s)) {
-          mlir::Value lenParam = operands[1]; // byte offset
-          auto bc =
-              rewriter.create<mlir::LLVM::BitcastOp>(loc, voidPtrTy(), base);
-          auto gep = genGEP(loc, ty, rewriter, bc, lenParam);
-          rewriter.replaceOpWithNewOp<mlir::LLVM::BitcastOp>(coor, ty, gep);
-          return success();
-        }
+    // Boxed type - get the base pointer from the box
+    if (baseObjectTy.dyn_cast<fir::BoxType>())
+      return doRewriteBox(coor, ty, operands, loc, rewriter);
+
+    // Reference or pointer type
+    if (baseObjectTy.isa<fir::ReferenceType, fir::PointerType>())
+      return doRewriteRefOrPtr(coor, ty, operands, loc, rewriter);
+
+    return rewriter.notifyMatchFailure(
+        coor, "fir.coordinate_of base operand has unsupported type");
+  }
+
+  unsigned getFieldNumber(fir::RecordType ty, mlir::Value op) const {
+    return fir::hasDynamicSize(ty)
+               ? op.getDefiningOp()
+                     ->getAttrOfType<mlir::IntegerAttr>("field")
+                     .getInt()
+               : getIntValue(op);
+  }
+
+  int64_t getIntValue(mlir::Value val) const {
+    assert(val && val.dyn_cast<mlir::OpResult>() && "must not be null value");
+    mlir::Operation *defop = val.getDefiningOp();
+
+    if (auto constOp = dyn_cast<mlir::arith::ConstantIntOp>(defop))
+      return constOp.value();
+    if (auto llConstOp = dyn_cast<mlir::LLVM::ConstantOp>(defop))
+      if (auto attr = llConstOp.value().dyn_cast<mlir::IntegerAttr>())
+        return attr.getValue().getSExtValue();
+    fir::emitFatalError(val.getLoc(), "must be a constant");
+  }
+
+  bool hasSubDimensions(mlir::Type type) const {
```

Comment:
Looks like this could be `static` instead of `const` and avoid passing `this` completely. (Possibly some of the other functions here as well.)

Reply:
Good point, I'll update it (as well as the other methods).